File: skills/skill-audit-ts/scripts/skill_lib.ts
// agent-tool: false
/**
 * Shared helpers for skill-audit-ts scripts.
 *
 * Imported by validate_skill.ts, analyze_skill.ts, recommend_scripts.ts, and others.
 * Node built-ins only.
 */

export type FrontmatterValue = string | Record<string, string>;
export type Frontmatter = Record<string, FrontmatterValue>;

// Terminal escape sequences (ESC-initiated). Catches the common forms; any
// stray ESC byte the regex misses still gets escaped to `\x1b` by the
// per-char fallback below, so this is about output cleanliness rather than
// security.
const ANSI_RE = new RegExp(
  "\\x1b(?:" +
    "\\[[\\x30-\\x3f]*[\\x20-\\x2f]*[\\x40-\\x7e]" + // CSI
    "|\\][^\\x07\\x1b]*(?:\\x07|\\x1b\\\\)" + // OSC (BEL or ST terminator)
    "|[PX^_][^\\x1b]*\\x1b\\\\" + // DCS / SOS / PM / APC
    "|[NO]" + // SS2 / SS3
    "|[\\x40-\\x5f]" + // other single-byte escapes
    ")",
  "g",
);

// Unicode codepoints that aren't visible but can manipulate or hide content.
// Escaped (not stripped) so they remain inspectable in the sanitized output.
export const DANGEROUS_UNICODE = new Set([
  // Bidi overrides + isolates (Trojan Source, RLO file-name attacks).
  "‪",
  "‫",
  "‬",
  "‭",
  "‮",
  "⁦",
  "⁧",
  "⁨",
  "⁩",
  // Invisible width / spacing. ZWJ (U+200D) is intentionally NOT here — it's
  // load-bearing inside emoji sequences.
  "​",
  "‌",
  "⁠",
  "﻿",
]);

const KEY_RE = /^([A-Za-z0-9_-]+):\s*(.*)$/;
const NESTED_RE = /^\s+([A-Za-z0-9_-]+):\s*(.*)$/;
const FOLDED_MARKERS = new Set([">", "|", ">-", "|-", ">+", "|+"]);
const FENCE_END_RE = /\r?\n---(?:\r?\n|$)/;

/**
 * Return [frontmatterObject, body] or [null, fullText] if no frontmatter.
 *
 * Handles a small subset of YAML sufficient for SKILL.md frontmatter: scalar
 * key/value, single-level nesting, folded (>) and block (|) scalars. Quoted
 * scalars have their quotes stripped.
 *
 * Strict on fence shape: the opening must be `---` followed by a newline, and
 * the closing must be `---` on its own line (or at end of file). A body line
 * beginning with `--- something` is body, not a closing fence.
 */
export function parseFrontmatter(text: string): [Frontmatter | null, string] {
  if (!text.startsWith("---")) {
    return [null, text];
  }
  if (text.length < 4 || (text[3] !== "\n" && text[3] !== "\r")) {
    // Opening fence must be on a line by itself.
    return [null, text];
  }
  const match = findFenceEnd(text, 3);
  if (!match) {
    return [null, text];
  }
  const fmText = text.slice(3, match.start).replace(/^[\r\n]+|[\r\n]+$/g, "");
  const body = text.slice(match.end);
  return [parseSimpleYaml(fmText), body];
}

/** Search for the closing `---` fence starting at `from`, in original-text coordinates. */
function findFenceEnd(text: string, from: number): { start: number; end: number } | null {
  const sub = text.slice(from);
  const m = FENCE_END_RE.exec(sub);
  if (!m) return null;
  return { start: from + m.index, end: from + m.index + m[0].length };
}

function collectIndented(lines: string[], start: number): [string[], number] {
  const collected: string[] = [];
  let i = start;
  while (i < lines.length && (lines[i]!.startsWith("  ") || lines[i]!.trim() === "")) {
    if (lines[i]!.trim() !== "") {
      collected.push(lines[i]!);
    }
    i += 1;
  }
  return [collected, i];
}

function parseSimpleYaml(text: string): Frontmatter {
  const result: Frontmatter = {};
  const lines = text.split(/\r?\n/);
  let i = 0;
  while (i < lines.length) {
    const line = lines[i]!;
    if (line.trim() === "" || line.trimStart().startsWith("#")) {
      i += 1;
      continue;
    }
    const m = KEY_RE.exec(line);
    if (!m) {
      i += 1;
      continue;
    }
    const key = m[1]!;
    const value = m[2]!.trimEnd();
    if (FOLDED_MARKERS.has(value)) {
      const [indented, next] = collectIndented(lines, i + 1);
      i = next;
      const sep = value.startsWith(">") ? " " : "\n";
      result[key] = indented.map((ln) => ln.trim()).join(sep);
      continue;
    }
    if (value === "") {
      const nested: Record<string, string> = {};
      const [indented, next] = collectIndented(lines, i + 1);
      i = next;
      for (const ln of indented) {
        const nm = NESTED_RE.exec(ln);
        if (nm) {
          nested[nm[1]!] = stripQuotes(nm[2]!.trim());
        }
      }
      result[key] = Object.keys(nested).length > 0 ? nested : "";
      continue;
    }
    result[key] = stripQuotes(value);
    i += 1;
  }
  return result;
}

function stripQuotes(value: string): string {
  if (
    value.length >= 2 &&
    ((value[0] === '"' && value[value.length - 1] === '"') ||
      (value[0] === "'" && value[value.length - 1] === "'"))
  ) {
    return value.slice(1, -1);
  }
  return value;
}

/**
 * Make value safe to print or include in JSON output.
 *
 * Untrusted SKILL.md content may contain ANSI escapes, control characters, or
 * unusual unicode that could confuse a downstream agent or terminal.
 *
 * This pass strips terminal escape sequences (CSI, OSC, DCS, single-byte
 * ESC), escapes C0 control characters as `\xNN`, and escapes a small set of
 * invisible / bidi-override Unicode codepoints (incl. U+202E "RLO" used in
 * Trojan Source attacks) as `\uXXXX`. Newline and tab pass through. ZWJ
 * (U+200D) passes through to keep emoji sequences intact. Length is capped
 * at maxLen.
 *
 * Not designed to be cryptographic. Don't pipe adversarial input to a
 * destination where stricter sanitization matters.
 */
export function sanitizeForEcho(value: unknown, maxLen = 200): string {
  let str = typeof value === "string" ? value : String(value);
  str = str.replace(ANSI_RE, "");
  const cleaned: string[] = [];
  for (const ch of str) {
    const cp = ch.codePointAt(0)!;
    if (ch === "\n" || ch === "\t") {
      cleaned.push(ch);
    } else if (cp < 0x20 || cp === 0x7f) {
      cleaned.push(`\\x${cp.toString(16).padStart(2, "0")}`);
    } else if (DANGEROUS_UNICODE.has(ch)) {
      cleaned.push(`\\u${cp.toString(16).padStart(4, "0")}`);
    } else {
      cleaned.push(ch);
    }
  }
  let out = cleaned.join("");
  if (out.length > maxLen) {
    out = out.slice(0, maxLen - 1) + "…";
  }
  return out;
}

/**
 * Emit structured error JSON to stderr.
 *
 * Provides a consistent shape for agent consumption:
 * {"error": <message>, "code": <dotted-code>, "hint": <recovery suggestion>}
 */
export function emitError(
  _script: string,
  message: string,
  opts: { code?: string; hint?: string } = {},
): void {
  const payload: Record<string, string> = { error: sanitizeForEcho(message, 500) };
  if (opts.code) payload.code = opts.code;
  if (opts.hint) payload.hint = opts.hint;
  process.stderr.write(JSON.stringify(payload) + "\n");
}

File: skills/skill-audit/scripts/skill_lib.py
"""Shared helpers for skill-audit scripts.

Imported by validate_skill.py, analyze_skill.py, and recommend_scripts.py.
Stdlib only.
"""
# agent-tool: false

from __future__ import annotations

import json
import re
import sys
from typing import Any

# Terminal escape sequences (ESC-initiated). Catches the common forms;
# any stray ESC byte the regex misses still gets escaped to `\x1b` by the
# per-char fallback in sanitize_for_echo, so this is about output cleanliness
# rather than security.
_ANSI_RE = re.compile(
    r"\x1b"
    r"(?:"
    r"\[[\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e]"  # CSI
    r"|\][^\x07\x1b]*(?:\x07|\x1b\\)"  # OSC (BEL or ST terminator)
    r"|[PX^_][^\x1b]*\x1b\\"  # DCS / SOS / PM / APC
    r"|[NO]"  # SS2 / SS3
    r"|[\x40-\x5f]"  # other single-byte escapes
    r")"
)

# Unicode codepoints that aren't visible but can manipulate or hide content.
# Escaped (not stripped) so they remain inspectable in the sanitized output.
DANGEROUS_UNICODE = frozenset(
    {
        # Bidi overrides + isolates (Trojan Source, RLO file-name attacks).
        "‪",
        "‫",
        "‬",
        "‭",
        "‮",
        "⁦",
        "⁧",
        "⁨",
        "⁩",
        # Invisible width / spacing. ZWJ (U+200D) is intentionally NOT here —
        # it's load-bearing inside emoji sequences.
        "\u200b",
        "‌",
        "⁠",
        "﻿",
    }
)

_KEY_RE = re.compile(r"^([A-Za-z0-9_-]+):\s*(.*)$")
_NESTED_RE = re.compile(r"^\s+([A-Za-z0-9_-]+):\s*(.*)$")
_FOLDED_MARKERS = (">", "|", ">-", "|-", ">+", "|+")
_FENCE_END_RE = re.compile(r"\r?\n---(?:\r?\n|\Z)")


def parse_frontmatter(text: str) -> tuple[dict[str, Any] | None, str]:
    """Return (frontmatter_dict, body) or (None, full_text) if no frontmatter.

    Handles a small subset of YAML sufficient for SKILL.md frontmatter:
    scalar key/value, single-level nesting, folded (>) and block (|) scalars.
    Quoted scalars have their quotes stripped.

    Strict on fence shape: the opening must be `---` followed by a newline,
    and the closing must be `---` on its own line (or at end of file). A
    body line beginning with `--- something` is body, not a closing fence.
    """
    if not text.startswith("---"):
        return None, text
    if len(text) < 4 or text[3] not in ("\n", "\r"):
        # Opening fence must be on a line by itself.
        return None, text
    match = _FENCE_END_RE.search(text, 3)
    if not match:
        return None, text
    fm_text = text[3 : match.start()].strip("\r\n")
    body = text[match.end() :]
    return _parse_simple_yaml(fm_text), body


def _collect_indented(lines: list[str], start: int) -> tuple[list[str], int]:
    """Return (non-empty indented lines, next index) from *start*."""
    collected: list[str] = []
    i = start
    while i < len(lines) and (lines[i].startswith("  ") or not lines[i].strip()):
        if lines[i].strip():
            collected.append(lines[i])
        i += 1
    return collected, i


def _parse_simple_yaml(text: str) -> dict[str, Any]:
    result: dict[str, Any] = {}
    lines = text.splitlines()
    i = 0
    while i < len(lines):
        line = lines[i]
        if not line.strip() or line.lstrip().startswith("#"):
            i += 1
            continue
        m = _KEY_RE.match(line)
        if not m:
            i += 1
            continue
        key, value = m.group(1), m.group(2).rstrip()
        if value in _FOLDED_MARKERS:
            indented, i = _collect_indented(lines, i + 1)
            sep = " " if value.startswith(">") else "\n"
            result[key] = sep.join(ln.strip() for ln in indented)
            continue
        if value == "":
            nested: dict[str, str] = {}
            indented, i = _collect_indented(lines, i + 1)
            for ln in indented:
                nm = _NESTED_RE.match(ln)
                if nm:
                    nested[nm.group(1)] = _strip_quotes(nm.group(2).strip())
            result[key] = nested or ""
            continue
        result[key] = _strip_quotes(value)
        i += 1
    return result


def _strip_quotes(value: str) -> str:
    if len(value) >= 2 and (
        (value[0] == '"' and value[-1] == '"') or (value[0] == "'" and value[-1] == "'")
    ):
        return value[1:-1]
    return value


def sanitize_for_echo(value: Any, max_len: int = 200) -> str:
    """Make value safe to print or include in JSON output.

    Untrusted SKILL.md content may contain ANSI escapes, control characters,
    or unusual unicode that could confuse a downstream agent or terminal.

    This pass strips terminal escape sequences (CSI, OSC, DCS, single-byte
    ESC), escapes C0 control characters as `\\xNN`, and escapes a small set
    of invisible / bidi-override Unicode codepoints (incl. U+202E "RLO" used
    in Trojan Source attacks) as `\\uXXXX`. Newline and tab pass through.
    ZWJ (U+200D) passes through to keep emoji sequences intact. Length is
    capped at max_len.

    Stdlib only. Not designed to be cryptographic. Don't pipe adversarial
    input to a destination where stricter sanitization matters.
    """
    if not isinstance(value, str):
        value = str(value)
    value = _ANSI_RE.sub("", value)
    cleaned: list[str] = []
    for ch in value:
        cp = ord(ch)
        if ch in {"\n", "\t"}:
            cleaned.append(ch)
        elif cp < 0x20 or cp == 0x7F:
            cleaned.append(f"\\x{cp:02x}")
        elif ch in DANGEROUS_UNICODE:
            cleaned.append(f"\\u{cp:04x}")
        else:
            cleaned.append(ch)
    out = "".join(cleaned)
    if len(out) > max_len:
        out = out[: max_len - 1] + "…"
    return out


def emit_error(
    script: str, message: str, *, code: str | None = None, hint: str | None = None
) -> None:
    """Emit structured error JSON to stderr.

    Provides a consistent shape for agent consumption:
    {"error": <message>, "code": <dotted-code>, "hint": <recovery suggestion>}
    """
    payload: dict[str, str] = {"error": sanitize_for_echo(message, max_len=500)}
    if code:
        payload["code"] = code
    if hint:
        payload["hint"] = hint
    print(json.dumps(payload), file=sys.stderr)
